Fix is_monotonic to check continuity before checking monotonicity

jhaayushkumar opened the pull request:
#### References to other Issues or PRs
Fixes #28578

#### Brief description of what is fixed or changed
Added continuity checks to [monotonicity_helper()](cci:1://file:///Users/ayushkumarjha/Desktop/sympy/sympy/calculus/singularities.py:130:0-183:49) and [is_monotonic()](cci:1://file:///Users/ayushkumarjha/Desktop/sympy/sympy/calculus/singularities.py:362:0-430:62) functions. Previously, these functions only verified that the derivative had no roots in the interval, without checking if the function was continuous throughout. This caused functions with poles to be incorrectly classified as monotonic.

**Before this fix:**
```python
from sympy import Symbol, Interval, tan
from sympy.calculus.singularities import is_monotonic, is_increasing

x = Symbol('x')
is_monotonic(tan(x), Interval(0, 5), x)   # Returns: True (incorrect)
is_increasing(tan(x), Interval(0, 5), x)  # Returns: True (incorrect)
# tan(x) has poles at pi/2 ≈ 1.57 and 3*pi/2 ≈ 4.71, both within [0, 5]
```

<!-- BEGIN RELEASE NOTES -->
* calculus
  * Fixed [is_monotonic](cci:1://file:///Users/ayushkumarjha/Desktop/sympy/sympy/calculus/singularities.py:362:0-430:62) and related functions to verify continuity before checking monotonicity. Functions with poles or discontinuities now correctly return `False`.
<!-- END RELEASE NOTES -->

smichr commented:
I wonder if None should be used if the interval is not continuous and save False only for when it is not monotonic but continuous. The None would signal that within the right interval it could be continuous. I am not familiar enough to decide if this is a good idea or not.

Peter230655 commented:
I have been following this discussion for a while, interesting!
What is the definition of ``monotonic`` to be applied here? If $f(x)$ monotonic for $x \in (a, b) \leftrightarrow f(x_1) \leq f(x_2)$ whenever  $x_1 \leq x_2$ and $x_1, x_2 \in (a, b)$ then continuity does not matter, it seems to me.

or am I missing something?

jhaayushkumar commented:
> cf [this question](https://math.stackexchange.com/questions/3644976/does-monotonic-function-imply-continuity)

Right — the examples in that discussion make the key point clear: monotonicity
does not require continuity, so a discontinuity alone should not force `False`.
It simply means the derivative/root-based logic cannot certify monotonicity on
the whole interval.

So returning `None` in the presence of discontinuities is the correct semantic
choice. That avoids incorrectly classifying discontinuous monotone functions as
non-monotonic and keeps the behavior consistent with SymPy’s general approach to
undecidable cases.

jhaayushkumar commented:
> A function can be monotonic without being continuous e.g. `Heaviside(x)` is monotonically increasing.
> 
> SymPy functions that return True, False or None use None as an indicator of "unknown" rather than to signify some other property.

Thanks for the clarification — that helps.

You're right that monotonicity does not imply continuity, and we shouldn’t use None to mean “discontinuous”. In the assumptions system, None should only indicate “SymPy cannot determine the truth value”.

For the integration logic, the only branch that matters is the case where is_monotonic(expr) is True. So both False (not monotone) and None (unknown/undetermined) should simply prevent the monotone-specific path from being taken, regardless of whether the function is continuous or not.

Given that, the revised behavior would be:

• True  → monotone over the given interval  
• False → definitely not monotone  
• None  → monotonicity cannot be determined (not necessarily discontinuous)

I’ll update the PR to avoid interpreting None as indicating a discontinuity and instead treat it strictly as “unknown” in line with the assumptions system.

jhaayushkumar commented:
> Given the comment [here](https://github.com/sympy/sympy/pull/28590#issuecomment-3536459120), is your change correct?

@smichr - Yes, the change is correct given how [singularities()](cci:1://file:///Users/ayushkumarjha/Desktop/sympy/sympy/calculus/singularities.py:28:0-122:58) works. Let me explain:

**Investigation shows** [singularities()](cci:1://file:///Users/ayushkumarjha/Desktop/sympy/sympy/calculus/singularities.py:28:0-122:58) only detects **poles/vertical asymptotes**, NOT jump discontinuities:

```python
>>> from sympy import *
>>> x = Symbol('x', real=True)

>>> singularities(Heaviside(x), x)
EmptySet  # Jump discontinuity NOT detected

>>> singularities(tan(x), x)
Union(ImageSet(...pi/2...), ImageSet(...3*pi/2...))  # Poles ARE detected

>>> singularities(1/x, x)
{0}  # Pole IS detected
```


"Would you like me to revert the changes to [is_monotonic](cci:1://file:///Users/ayushkumarjha/Desktop/sympy/sympy/calculus/singularities.py:366:0-438:62) and keep only the integration continuity checks?"

jhaayushkumar committed:
Fix is_monotonic to reject functions with interior poles (issue #28578)

Functions with poles cannot be monotonic per Rudin's Theorem 4.30. Added singularity
detection and comprehensive docstring with mathematical definitions. Tests now correctly
return False for tan(x) on [0,5] and similar cases with interior discontinuities.
